Invalid Trajectory: start point deviates

This reports problems with #63 in practice.

I just tested current `indigo-devel` with our PR2 and I often get these errors for the execution of trajectories (only one each time):

> [ERROR] [1476283190.409951701]: 
Invalid Trajectory: start point deviates from current robot state more than 0.01
joint 'l_forearm_roll_joint': expected: 2.62748, current: -3.6557

> [ERROR] [1476283258.133136672]: 
Invalid Trajectory: start point deviates from current robot state more than 0.01
joint 'l_forearm_roll_joint': expected: 0.469322, current: -5.81421

This arises from un-normalized joint values of the unlimited joints on the PR2's arm, that are not taken into account. This is a systematic problem, and imho not just specific to the PR2.

Also I think it's worth to note that it is still necessary to update the start state manually when planning in RViz, because the update happens too early after the execution (the rviz plugin does not have a current state of the robot yet). Without updating, I keep getting errors such as

> [ERROR] [1476283118.443709588]: 
Invalid Trajectory: start point deviates from current robot state more than 0.01
joint 'l_elbow_flex_joint': expected: -1.17855, current: -1.18984

> [ERROR] [1476287101.452005358]: 
Invalid Trajectory: start point deviates from current robot state more than 0.01
joint 'l_upper_arm_roll_joint': expected: 3.67128, current: 3.69101

If we should decide to release into indigo without #232 I consider it reasonable to add a sleep to the plugin until this is fixed properly by #232.

As the first type of normalization error breaks working setups (at least the PR2), I consider this a blocker for the indigo as well as the kinetic release.

I'm really sorry to keep on adding release blockers. :(

--- a/moveit_ros/planning/trajectory_execution_manager/src/trajectory_execution_manager.cpp
+++ b/moveit_ros/planning/trajectory_execution_manager/src/trajectory_execution_manager.cpp
@@ -896,6 +896,34 @@ bool TrajectoryExecutionManager::validate(const TrajectoryExecutionContext &cont
     ROS_WARN_NAMED("traj_execution", "Failed to validate trajectory: couldn't receive full current joint state within 1s");
     return false;
   }
+  current_state->enforceBounds();
+
+  // build a RobotState that can be used to compare current_state and the start state with enforced bounds
+  robot_state::RobotState bounded_start_state(*current_state);
+  for (std::vector<moveit_msgs::RobotTrajectory>::const_iterator traj_it = context.trajectory_parts_.begin();
+       traj_it != context.trajectory_parts_.end(); ++traj_it)
+  {
+    const std::vector<double> &positions = traj_it->joint_trajectory.points.front().positions;
+    const std::vector<std::string> &joint_names = traj_it->joint_trajectory.joint_names;
+    const std::size_t n = joint_names.size();
+
+    if (positions.size() != n)
+    {
+      ROS_ERROR_NAMED("traj_execution", "Wrong trajectory: #joints: %zu != #positions: %zu", n, positions.size());
+      return false;
+    }
+
+    try
+    {
+      bounded_start_state.setVariablePositions(joint_names, positions);
+    }
+    catch(moveit::Exception& e)
+    {
+      ROS_ERROR_STREAM_NAMED("traj_execution", "Unknown joint in trajectory: " << e.what());
+      return false;
+    }
+  }
+  bounded_start_state.enforceBounds();
 
   for (std::vector<moveit_msgs::RobotTrajectory>::const_iterator traj_it = context.trajectory_parts_.begin();
        traj_it != context.trajectory_parts_.end(); ++traj_it)
@@ -903,22 +931,14 @@ bool TrajectoryExecutionManager::validate(const TrajectoryExecutionContext &cont
     const std::vector<double> &positions = traj_it->joint_trajectory.points.front().positions;
     const std::vector<std::string> &joint_names = traj_it->joint_trajectory.joint_names;
     const std::size_t n = joint_names.size();
-    if (positions.size() != n)
-    {
-      ROS_ERROR_NAMED("traj_execution", "Wrong trajectory: #joints: %zu != #positions: %zu", n, positions.size());
-      return false;
-    }
 
     for (std::size_t i = 0; i < n; ++i)
     {
       const robot_model::JointModel *jm = current_state->getJointModel(joint_names[i]);
-      if (!jm)
-      {
-        ROS_ERROR_STREAM_NAMED("traj_execution", "Unknown joint in trajectory: " << joint_names[i]);
-        return false;
-      }
+      const robot_model::JointModel *jm_start = bounded_start_state.getJointModel(joint_names[i]);
+
       // TODO: check multi-DoF joints ?
-      if (fabs(current_state->getJointPositions(jm)[0] - positions[i]) > allowed_start_tolerance_)
+      if (fabs(current_state->getJointPositions(jm)[0] - bounded_start_state.getJointPositions(jm_start)[0]) > allowed_start_tolerance_)
       {
         ROS_ERROR_NAMED("traj_execution",
                         "\nInvalid Trajectory: start point deviates from current robot state more than %g"

--- a/moveit_ros/planning/trajectory_execution_manager/src/trajectory_execution_manager.cpp
+++ b/moveit_ros/planning/trajectory_execution_manager/src/trajectory_execution_manager.cpp
@@ -896,55 +896,40 @@ bool TrajectoryExecutionManager::validate(const TrajectoryExecutionContext &cont
     ROS_WARN_NAMED("traj_execution", "Failed to validate trajectory: couldn't receive full current joint state within 1s");
     return false;
   }
-  current_state->enforceBounds();
 
-  // build a RobotState that can be used to compare current_state and the start state with enforced bounds
-  robot_state::RobotState bounded_start_state(*current_state);
   for (std::vector<moveit_msgs::RobotTrajectory>::const_iterator traj_it = context.trajectory_parts_.begin();
        traj_it != context.trajectory_parts_.end(); ++traj_it)
   {
     const std::vector<double> &positions = traj_it->joint_trajectory.points.front().positions;
     const std::vector<std::string> &joint_names = traj_it->joint_trajectory.joint_names;
     const std::size_t n = joint_names.size();
-
     if (positions.size() != n)
     {
       ROS_ERROR_NAMED("traj_execution", "Wrong trajectory: #joints: %zu != #positions: %zu", n, positions.size());
       return false;
     }
 
-    try
-    {
-      bounded_start_state.setVariablePositions(joint_names, positions);
-    }
-    catch(moveit::Exception& e)
-    {
-      ROS_ERROR_STREAM_NAMED("traj_execution", "Unknown joint in trajectory: " << e.what());
-      return false;
-    }
-  }
-  bounded_start_state.enforceBounds();
-
-  for (std::vector<moveit_msgs::RobotTrajectory>::const_iterator traj_it = context.trajectory_parts_.begin();
-       traj_it != context.trajectory_parts_.end(); ++traj_it)
-  {
-    const std::vector<double> &positions = traj_it->joint_trajectory.points.front().positions;
-    const std::vector<std::string> &joint_names = traj_it->joint_trajectory.joint_names;
-    const std::size_t n = joint_names.size();
-
     for (std::size_t i = 0; i < n; ++i)
     {
       const robot_model::JointModel *jm = current_state->getJointModel(joint_names[i]);
-      const robot_model::JointModel *jm_start = bounded_start_state.getJointModel(joint_names[i]);
+      if (!jm)
+      {
+        ROS_ERROR_STREAM_NAMED("traj_execution", "Unknown joint in trajectory: " << joint_names[i]);
+        return false;
+      }
 
       // TODO: check multi-DoF joints ?
-      if (fabs(current_state->getJointPositions(jm)[0] - bounded_start_state.getJointPositions(jm_start)[0]) > allowed_start_tolerance_)
+      double cur_position = current_state->getJointPositions(jm)[0];
+      double traj_position = positions[i];
+      // normalize positions and compare
+      jm->enforcePositionBounds(&cur_position);
+      jm->enforcePositionBounds(&traj_position);
+      if (fabs(cur_position - traj_position) > allowed_start_tolerance_)
       {
         ROS_ERROR_NAMED("traj_execution",
                         "\nInvalid Trajectory: start point deviates from current robot state more than %g"
                         "\njoint '%s': expected: %g, current: %g",
-                        allowed_start_tolerance_,
-                        joint_names[i].c_str(), positions[i], current_state->getJointPositions(jm)[0]);
+                        allowed_start_tolerance_, joint_names[i].c_str(), traj_position, cur_position);
         return false;
       }
     }

